### PR TITLE
gRPC Change

### PR DESCRIPTION
**Pull request by anfernee**
Tested @mikedanese suggestion. We could use metadata for the remote address, then we could use 1:1 mapping between proxied connection and grpc stream, which made the grpc interface very simple.

### CODE REVIEW

File: keps/sig-api-machinery/20190226-network-proxy.md [hunk 1]
```diff
@@ -241,102 +241,29 @@ or which user/tenant tries to initiate the request, etc.
...
-    // be IP address.
-    string address = 2;
-
-    // random is the randomly generated bytes that represents a given dial request.
-    // the number is kept unchanged across the proxies, and copied to DialResponse
-    // from the other side of the tunnel. The number cannot be reused across 
-    // different DialRequests.
-    bytes random = 3;
+// ConnectParam defines the remote TCP endpoint to connect
+message ConnectParam {
```
Comment (mikedanese) [inline keps/sig-api-machinery/20190226-network-proxy.md]: I suggest calling this ConnectOptions

File: keps/sig-api-machinery/20190226-network-proxy.md [hunk 2]
```diff
@@ -241,102 +241,29 @@ or which user/tenant tries to initiate the request, etc.
...
-  EOF = 0;
-  EIO = 1;
-  ECONNRESET = 2;
-  ETIMEOUT = 3;
-  EADDRNOTAVAIL = 4;
-  EMFILE = 5;
-  // … 
+  // Proxy a TCP connection to a remote address defined by ConnectParam.
+  // The ConnectParam is defined in metadata under key "x-kube-proxy/v1".
+  // metadata["x-kube-proxy/v1"] = base64.Encode(proto.Marshal(connectParam))
```
Comment (mikedanese) [inline keps/sig-api-machinery/20190226-network-proxy.md]: Couple things:
1) base64 is not necessary over grpc or http2
2) "kube-proxy" also means something very different than what you are doing here
3) Why would version the header vs versioning the entire service?

**Review by mikedanese — COMMENTED** [review]
This is awesome! Thanks!

File: keps/sig-api-machinery/20190226-network-proxy.md (hunk 2 above)
Comment (anfernee) [reply keps/sig-api-machinery/20190226-network-proxy.md]: 2-3) +1. version will cause confusing. I will remove it. 
1) I want the metadata value is really a valid string.

File: keps/sig-api-machinery/20190226-network-proxy.md (hunk 1 above)
Comment (anfernee) [reply keps/sig-api-machinery/20190226-network-proxy.md]: +1

**Comment by mikedanese** [comment]
LGTM, thanks.

**Review by cheftako — APPROVED** [review]
/lgtm

**Comment by anfernee** [comment]
Thanks @cheftako . merging this then the change will show up in https://github.com/kubernetes/enhancements/pull/872/commits we could squash into one commit later.

**Comment by anfernee** [comment]
Only you could merge it @cheftako  `Only those with write access to this repository can merge pull requests.`

**Pull request merged**